airflow/jobs/base_job_runner.py

```diff
@@ -17,28 +17,31 @@
 
 from __future__ import annotations
 
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Generic, TypeVar
 
 from airflow.utils.session import NEW_SESSION, provide_session
 
 if TYPE_CHECKING:
     from sqlalchemy.orm import Session
 
     from airflow.jobs.job import Job
+    from airflow.serialization.pydantic.job import JobPydantic
 
+JobOrJobPydantic = TypeVar("JobOrJobPydantic", Job, JobPydantic)
 
-class BaseJobRunner:
+
+class BaseJobRunner(Generic[JobOrJobPydantic]):
     """Abstract class for job runners to derive from."""
 
     job_type = "undefined"
 
-    def __init__(self, job):
+    def __init__(self, job: JobOrJobPydantic):
         if job.job_type and job.job_type != self.job_type:
             raise Exception(
                 f"The job is already assigned a different job_type: {job.job_type}."
                 f"This is a bug and should be reported."
             )
-        self.job = job
+        self.job: JobOrJobPydantic = job
```

Comment:
I don’t think this annotation is needed?

Reply:
MyPy complained when it was not there - not sure why

Reply:
MyPy complained hen it wasn't there. But just wondering - isn't the original approach with both  T and JobOrJobPydantic needed actually? The first is the generic type (which can also be just Job), and the other actually a type definition?

Reply:
> The first is the generic type (which can also be just Job), and the other actually a type definition

A TypeVar is a, well, variable (name), and that variable name can be used in different contexts. One name can serve these two purposes without conflicting internally, so it works fine.

Reply:
> MyPy complained when it was not there - not sure why

What’s the error message? Maybe that can be categorised as a deficit and submitted to Mypy for fixing.

Reply:
Pretty straightforward :)

Reply:
> A TypeVar is a, well, variable (name), and that variable name can be used in different contexts. One name can serve these two purposes without conflicting internally, so it works fine.

I see  - but I think (i just re-added it). it's a bit clearer and more explicit now what I separated them into two vars - It would be rather confusing (pairing with the mypy deficit) to see the job declared with JobOrPydantic, where it can be just Job actually. I thikn having T makes it  a bit more clear what the intention is here.

Reply:
BTW. This is what I do not like too much about the typing system's "flexibility" in Python (especially when it is abused) - same things can be used in different context with different meaning and sometimes when I look at the code that has them mixed, it's really hard to wrap your head around it.